storm-core/src/jvm/org/apache/storm/Thrift.java

```diff
@@ -0,0 +1,351 @@
+import org.apache.storm.generated.ComponentObject;
+
+import org.apache.storm.task.IBolt;
+import org.apache.storm.topology.BoltDeclarer;
+import org.apache.storm.topology.IRichBolt;
+import org.apache.storm.topology.IBasicBolt;
+import org.apache.storm.topology.IRichSpout;
+import org.apache.storm.topology.SpoutDeclarer;
+import org.json.simple.JSONValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.storm.utils.Utils;
+import org.apache.storm.grouping.CustomStreamGrouping;
+import org.apache.storm.topology.TopologyBuilder;
+
+public class Thrift {
+    private static Logger LOG = LoggerFactory.getLogger(Thrift.class);
+
+    private static StormTopology._Fields[] STORM_TOPOLOGY_FIELDS = null;
+    private static StormTopology._Fields[] SPOUT_FIELDS =
+            { StormTopology._Fields.SPOUTS, StormTopology._Fields.STATE_SPOUTS };
+
+    static {
+        Set<_Fields> keys = StormTopology.metaDataMap.keySet();
+        keys.toArray(STORM_TOPOLOGY_FIELDS = new StormTopology._Fields[keys.size()]);
+    }
+
+    public static StormTopology._Fields[] getTopologyFields() {
+        return STORM_TOPOLOGY_FIELDS;
+    }
+
+    public static StormTopology._Fields[] getSpoutFields() {
+        return SPOUT_FIELDS;
+    }
+
+    public static class SpoutDetails {
+        private IRichSpout spout;
+        private Integer parallelism;
+        private Map conf;
+
+        public SpoutDetails(IRichSpout spout, Integer parallelism, Map conf) {
+            this.spout = spout;
+            this.parallelism = parallelism;
+            this.conf = conf;
+        }
+
+        public IRichSpout getSpout() {
+            return spout;
+        }
+
+        public Integer getParallelism() {
+            return parallelism;
+        }
+
+        public Map getConf() {
+            return conf;
+        }
+    }
+
+    public static class BoltDetails {
+        private Object bolt;
+        private Map conf;
+        private Integer parallelism;
+        private Map<GlobalStreamId, Grouping> inputs;
+
+        public BoltDetails(Object bolt, Map conf, Integer parallelism,
+                           Map<GlobalStreamId, Grouping> inputs) {
+            this.bolt = bolt;
+            this.conf = conf;
+            this.parallelism = parallelism;
+            this.inputs = inputs;
+        }
+
+        public Object getBolt() {
+            return bolt;
+        }
+
+        public Map getConf() {
+            return conf;
+        }
+
+        public Map<GlobalStreamId, Grouping> getInputs() {
+            return inputs;
+        }
+
+        public Integer getParallelism() {
+            return parallelism;
+        }
+    }
+
+    public static StreamInfo directOutputFields(List<String> fields) {
+        return new StreamInfo(fields, true);
+    }
+
+    public static StreamInfo outputFields(List<String> fields) {
+        return new StreamInfo(fields, false);
+    }
+
+    public static Grouping prepareShuffleGrouping() {
```

Comment:
I'm not sure I care too much but the original function was mk-shuffle-grouping not prepare-shuffle-grouping